src/Sulu/Bundle/SecurityBundle/User/UserProvider.php

```diff
@@ -12,25 +12,29 @@
 namespace Sulu\Bundle\SecurityBundle\User;
 
 use Doctrine\ORM\NoResultException;
+use Doctrine\Persistence\ManagerRegistry;
 use Sulu\Bundle\SecurityBundle\System\SystemStoreInterface;
 use Sulu\Component\Security\Authentication\UserInterface;
 use Sulu\Component\Security\Authentication\UserRepositoryInterface;
 use Symfony\Component\Security\Core\Exception\DisabledException;
 use Symfony\Component\Security\Core\Exception\LockedException;
 use Symfony\Component\Security\Core\Exception\UnsupportedUserException;
 use Symfony\Component\Security\Core\Exception\UserNotFoundException;
+use Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface;
+use Symfony\Component\Security\Core\User\PasswordUpgraderInterface;
 use Symfony\Component\Security\Core\User\UserInterface as BaseUserInterface;
 use Symfony\Component\Security\Core\User\UserProviderInterface;
 
 /**
  * Responsible for loading the user from the database for the Symfony security system. Takes also the security system
  * configuration from the webspaces into account.
  */
-class UserProvider implements UserProviderInterface
+class UserProvider implements UserProviderInterface, PasswordUpgraderInterface
 {
     public function __construct(
         private UserRepositoryInterface $userRepository,
         private SystemStoreInterface $systemStore,
+        private ManagerRegistry $doctrine,
     ) {
     }
```

Comment:
We normally directly inject our entity manager, for BC reasons it still need to be nullable:

```suggestion
        private ?EntityManagerInterface $entityManager,
    ) {
        if (null === $this->entityManager) {
            @trigger_deprecation(
                'sulu/sulu',
                '2.5',
                \sprintf(
                    'The usage of the "%s" without setting "$entityManager" is deprecated and will not longer work in Sulu 3.0.',
                    self::class
                )
            );
        }
    }
```